<a href="https://colab.research.google.com/github/AryanShukla-1/My_Projects/blob/main/jupyter_labs_eda_sql_coursera_sqllite.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>

<h1 align=center><font size = 5>Assignment: SQL Notebook for Peer Assignment</font></h1>

Estimated time needed: **60** minutes.

## Introduction
Using this Python notebook you will:

1.  Understand the Spacex DataSet
2.  Load the dataset  into the corresponding table in a Db2 database
3.  Execute SQL queries to answer assignment questions


## Overview of the DataSet

SpaceX has gained worldwide attention for a series of historic milestones.

It is the only private company ever to return a spacecraft from low-earth orbit, which it first accomplished in December 2010.
SpaceX advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars wheras other providers cost upward of 165 million dollars each, much of the savings is because Space X can reuse the first stage.


Therefore if we can determine if the first stage will land, we can determine the cost of a launch.

This information can be used if an alternate company wants to bid against SpaceX for a rocket launch.

This dataset includes a record for each payload carried during a SpaceX mission into outer space.


### Download the datasets

This assignment requires you to load the spacex dataset.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. Click on the link below to download and save the dataset (.CSV file):

 <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/data/Spacex.csv" target="_blank">Spacex DataSet</a>



In [ ]:
!pip install sqlalchemy==1.3.9


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 81.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for sqlalchemy: filename=SQLAlchemy-1.3.9-cp312-cp312-linux_x86_64.whl size=1196052 sha256=2e273235e7b719b59ad6108e2eaa85b6a7eb674fd0e8c10339dcd82417291033
  Stored in directory: /root/.cache/pip/wheels/b3/1c/42/0e26b8d512adc6bce10ff71a05229366b4ccec641cd3b42111
Successfully built sqlalchemy
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 2.0.44
    Uninstalling SQLAlchemy-2.0.44:
      Successfully uninstalled SQLAlchemy-2.0.44
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
alembic 1.17.2 requires SQLAlchemy>=1.4.0, but you have sqlalchemy 1.3.9 which is incompatible.
ipython-sql 0.5.0 requires sqlalchemy>=2.0, but you have sqlalchemy 1.3.9 which is incompatible.
google-adk 1.19.0 requires sqlal

### Connect to the database

Let us first load the SQL extension and establish a connection with the database


In [ ]:
!pip install ipython-sql
!pip install ipython-sql prettytable

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 56.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 52.2 MB/s eta 0:00:00
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 1.3.9
    Uninstalling SQLAlchemy-1.3.9:
      Successfully uninstalled SQLAlchemy-1.3.9


In [ ]:
%load_ext sql

In [ ]:
import csv, sqlite3
import prettytable
prettytable.DEFAULT = 'DEFAULT'

con = sqlite3.connect("my_data1.db")
cur = con.cursor()

In [ ]:
!pip install -q pandas

In [ ]:
%sql sqlite:///my_data1.db

In [ ]:
import pandas as pd
df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/data/Spacex.csv")
df.to_sql("SPACEXTBL", con, if_exists='replace', index=False,method="multi")

101

**Note:This below code is added to remove blank rows from table**


In [ ]:
#DROP THE TABLE IF EXISTS

%sql DROP TABLE IF EXISTS SPACEXTABLE;

 * sqlite:///my_data1.db
Done.


[]

In [ ]:
%sql create table SPACEXTABLE as select * from SPACEXTBL where Date is not null

 * sqlite:///my_data1.db
Done.


[]

## Tasks

Now write and execute SQL queries to solve the assignment tasks.

**Note: If the column names are in mixed case enclose it in double quotes
   For Example "Landing_Outcome"**

### Task 1




##### Display the names of the unique launch sites  in the space mission



### Task 2


#####  Display 5 records where launch sites begin with the string 'CCA'


### Task 3




##### Display the total payload mass carried by boosters launched by NASA (CRS)


### Task 4




##### Display average payload mass carried by booster version F9 v1.1


### Task 5

##### List the date when the first succesful landing outcome in ground pad was acheived.


_Hint:Use min function_


### Task 6

##### List the names of the boosters which have success in drone ship and have payload mass greater than 4000 but less than 6000


### Task 7




##### List the total number of successful and failure mission outcomes


### Task 8



##### List all the booster_versions that have carried the maximum payload mass, using a subquery with a suitable aggregate function.


### Task 9


##### List the records which will display the month names, failure landing_outcomes in drone ship ,booster versions, launch_site for the months in year 2015.

**Note: SQLLite does not support monthnames. So you need to use  substr(Date, 6,2) as month to get the months and substr(Date,0,5)='2015' for year.**


### Task 10




##### Rank the count of landing outcomes (such as Failure (drone ship) or Success (ground pad)) between the date 2010-06-04 and 2017-03-20, in descending order.


### Reference Links

* <a href ="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Labs_Coursera_V5/labs/Lab%20-%20String%20Patterns%20-%20Sorting%20-%20Grouping/instructional-labs.md.html?origin=www.coursera.org">Hands-on Lab : String Patterns, Sorting and Grouping</a>  

*  <a  href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Labs_Coursera_V5/labs/Lab%20-%20Built-in%20functions%20/Hands-on_Lab__Built-in_Functions.md.html?origin=www.coursera.org">Hands-on Lab: Built-in functions</a>

*  <a  href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Labs_Coursera_V5/labs/Lab%20-%20Sub-queries%20and%20Nested%20SELECTs%20/instructional-labs.md.html?origin=www.coursera.org">Hands-on Lab : Sub-queries and Nested SELECT Statements</a>

*   <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Module%205/DB0201EN-Week3-1-3-SQLmagic.ipynb">Hands-on Tutorial: Accessing Databases with SQL magic</a>

*  <a href= "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/Module%205/DB0201EN-Week3-1-4-Analyzing.ipynb">Hands-on Lab: Analyzing a real World Data Set</a>




In [ ]:
import pandas as pd
import sqlite3

# Connect to the SQLite database
con = sqlite3.connect("my_data1.db")

# Load the SPACEXTABLE into a pandas DataFrame
df_spacex = pd.read_sql_query("SELECT * FROM SPACEXTABLE", con)

# Close the connection
con.close()

# Get the number of unique values for the specified columns
unique_orbits = df_spacex['Orbit'].nunique()
unique_landing_outcomes = df_spacex['Landing_Outcome'].nunique()
unique_booster_versions = df_spacex['Booster_Version'].nunique()

print(f"Number of unique values in 'Orbit': {unique_orbits}")
print(f"Number of unique values in 'Landing_Outcome': {unique_landing_outcomes}")
print(f"Number of unique values in 'Booster_Version': {unique_booster_versions}")

Number of unique values in 'Orbit': 8
Number of unique values in 'Landing_Outcome': 11
Number of unique values in 'Booster_Version': 97


## Author(s)

<h4> Lakshmi Holla </h4>


## Other Contributors

<h4> Rav Ahuja </h4>


<!--
## Change log
| Date | Version | Changed by | Change Description |
|------|--------|--------|---------|
| 2024-07-10 | 1.1 |Anita Verma | Changed Version|
| 2021-07-09 | 0.2 |Lakshmi Holla | Changes made in magic sql|
| 2021-05-20 | 0.1 |Lakshmi Holla | Created Initial Version |
-->


## <h3 align="center"> © IBM Corporation 2021. All rights reserved. <h3/>


# Task
Display unique launch sites, 5 records of launch sites starting with 'CCA', total payload mass for NASA (CRS), average payload mass for F9 v1.1, the date of the first successful ground pad landing, booster names with successful drone ship landing and payload mass between 4000 and 6000 kg, total successful and failure mission outcomes, booster versions with maximum payload mass, month names, failure drone ship landing outcomes, booster versions, and launch sites for 2015, and rank landing outcomes by count between 2010-06-04 and 2017-03-20. Finally, summarize the results and provide insights.

## Task 1: Display Unique Launch Sites

### Subtask:
Display the names of the unique launch sites in the space mission.


**Reasoning**:
The subtask is to display the unique launch sites. This can be achieved by executing a SQL query to select distinct values from the 'Launch_Site' column in the 'SPACEXTABLE'.



In [ ]:
%%sql
SELECT DISTINCT "Launch_Site" FROM SPACEXTABLE;

 * sqlite:///my_data1.db
Done.


[('CCAFS LC-40',), ('VAFB SLC-4E',), ('KSC LC-39A',), ('CCAFS SLC-40',)]

## Task 2: Display Launch Sites starting with 'CCA'

### Subtask:
Display 5 records where launch sites begin with the string 'CCA' by executing a SQL query to select 5 records from 'SPACEXTABLE' where 'Launch_Site' LIKE 'CCA%'.


**Reasoning**:
The subtask requires displaying 5 records where the launch site starts with 'CCA'. This can be achieved by writing a SQL query to select all columns from 'SPACEXTABLE', filtering for 'Launch_Site' like 'CCA%', and limiting the results to 5.



In [ ]:
%%sql
SELECT * FROM SPACEXTABLE
WHERE "Launch_Site" LIKE 'CCA%'
LIMIT 5;

 * sqlite:///my_data1.db
Done.


[('2010-06-04', '18:45:00', 'F9 v1.0  B0003', 'CCAFS LC-40', 'Dragon Spacecraft Qualification Unit', 0, 'LEO', 'SpaceX', 'Success', 'Failure (parachute)'),
 ('2010-12-08', '15:43:00', 'F9 v1.0  B0004', 'CCAFS LC-40', 'Dragon demo flight C1, two CubeSats, barrel of Brouere cheese', 0, 'LEO (ISS)', 'NASA (COTS) NRO', 'Success', 'Failure (parachute)'),
 ('2012-05-22', '7:44:00', 'F9 v1.0  B0005', 'CCAFS LC-40', 'Dragon demo flight C2', 525, 'LEO (ISS)', 'NASA (COTS)', 'Success', 'No attempt'),
 ('2012-10-08', '0:35:00', 'F9 v1.0  B0006', 'CCAFS LC-40', 'SpaceX CRS-1', 500, 'LEO (ISS)', 'NASA (CRS)', 'Success', 'No attempt'),
 ('2013-03-01', '15:10:00', 'F9 v1.0  B0007', 'CCAFS LC-40', 'SpaceX CRS-2', 677, 'LEO (ISS)', 'NASA (CRS)', 'Success', 'No attempt')]

**Reasoning**:
The subtask is to display the total payload mass carried by boosters launched by NASA (CRS). This can be achieved by summing the 'Payload_Mass__kg_' column from the 'SPACEXTABLE' where the 'Customer' is 'NASA (CRS)'.



In [ ]:
%%sql
SELECT SUM("Payload_Mass__kg_") FROM SPACEXTABLE WHERE Customer = 'NASA (CRS)';

 * sqlite:///my_data1.db
Done.


[(45596,)]

## Average Payload Mass for F9 v1.1

### Subtask:
Display the average payload mass carried by booster version F9 v1.1.


**Reasoning**:
The subtask is to display the average payload mass carried by booster version F9 v1.1. This can be achieved by executing a SQL query to calculate the average of the 'Payload_Mass__kg_' column from the 'SPACEXTABLE' where the 'Booster_Version' is 'F9 v1.1'.



In [ ]:
%%sql
SELECT AVG("Payload_Mass__kg_") FROM SPACEXTABLE WHERE "Booster_Version" = 'F9 v1.1';

 * sqlite:///my_data1.db
Done.


[(2928.4,)]

## Task 5: Date of First Successful Ground Pad Landing

### Subtask:
List the date when the first successful landing outcome in 'ground pad' was achieved.


**Reasoning**:
The subtask is to find the date of the first successful landing outcome on a 'ground pad'. This can be achieved by executing a SQL query to select the minimum 'Date' from the 'SPACEXTABLE' where the 'Landing_Outcome' is 'Success (ground pad)'.



In [ ]:
%%sql
SELECT MIN(Date) FROM SPACEXTABLE WHERE "Landing_Outcome" = 'Success (ground pad)';

 * sqlite:///my_data1.db
Done.


[('2015-12-22',)]

## Boosters with Successful Drone Ship Landing and Specific Payload Mass

### Subtask:
List the names of the boosters which have a 'Success (drone ship)' landing outcome and a payload mass greater than 4000 but less than 6000.


**Reasoning**:
The subtask requires listing booster versions based on specific landing outcomes and payload mass ranges. A SQL query is needed to select distinct booster versions from the 'SPACEXTABLE', filtering by 'Landing_Outcome' being 'Success (drone ship)' and 'Payload_Mass__kg_' being between 4000 and 6000.



In [ ]:
%%sql
SELECT DISTINCT "Booster_Version"
FROM SPACEXTABLE
WHERE "Landing_Outcome" = 'Success (drone ship)' AND "Payload_Mass__kg_" > 4000 AND "Payload_Mass__kg_" < 6000;

 * sqlite:///my_data1.db
Done.


[('F9 FT B1022',), ('F9 FT B1026',), ('F9 FT  B1021.2',), ('F9 FT  B1031.2',)]

## Total Successful and Failure Mission Outcomes

### Subtask:
List the total number of successful and failure mission outcomes.


**Reasoning**:
To list the total number of successful and failure mission outcomes, I need to count the occurrences of each unique mission outcome from the 'SPACEXTABLE' and group them by 'Mission_Outcome'.



In [ ]:
%%sql
SELECT "Mission_Outcome", COUNT("Mission_Outcome") AS Outcome_Count
FROM SPACEXTABLE
GROUP BY "Mission_Outcome";

 * sqlite:///my_data1.db
Done.


[('Failure (in flight)', 1),
 ('Success', 98),
 ('Success ', 1),
 ('Success (payload status unclear)', 1)]

## Task 8: Booster Versions with Maximum Payload Mass

### Subtask:
List all the booster_versions that have carried the maximum payload mass, using a subquery with a suitable aggregate function.


**Reasoning**:
The subtask is to list booster versions that carried the maximum payload mass. This can be achieved by using a subquery to first find the maximum payload mass, and then selecting the booster versions from the main table that match this maximum payload mass.



In [ ]:
%%sql
SELECT DISTINCT "Booster_Version"
FROM SPACEXTABLE
WHERE "Payload_Mass__kg_" = (
    SELECT MAX("Payload_Mass__kg_")
    FROM SPACEXTABLE
);

 * sqlite:///my_data1.db
Done.


[('F9 B5 B1048.4',),
 ('F9 B5 B1049.4',),
 ('F9 B5 B1051.3',),
 ('F9 B5 B1056.4',),
 ('F9 B5 B1048.5',),
 ('F9 B5 B1051.4',),
 ('F9 B5 B1049.5',),
 ('F9 B5 B1060.2 ',),
 ('F9 B5 B1058.3 ',),
 ('F9 B5 B1051.6',),
 ('F9 B5 B1060.3',),
 ('F9 B5 B1049.7 ',)]

**Reasoning**:
The next subtask is to list records showing month names, failure landing outcomes in drone ship, booster versions, and launch sites for the year 2015. This requires filtering by date and landing outcome, and extracting the month from the date. As SQLite does not support `MONTHNAME`, `substr(Date, 6,2)` will be used to extract the month, and `substr(Date,0,5)='2015'` to filter by year.



In [ ]:
%%sql
SELECT
  substr(Date, 6, 2) AS Month,
  "Landing_Outcome",
  "Booster_Version",
  "Launch_Site"
FROM SPACEXTABLE
WHERE
  substr(Date, 0, 5) = '2015' AND "Landing_Outcome" LIKE 'Failure (drone ship)%';

 * sqlite:///my_data1.db
Done.


[('01', 'Failure (drone ship)', 'F9 v1.1 B1012', 'CCAFS LC-40'),
 ('04', 'Failure (drone ship)', 'F9 v1.1 B1015', 'CCAFS LC-40')]

## Task 10: Rank Landing Outcomes by Count

### Subtask:
Rank the count of landing outcomes (such as Failure (drone ship) or Success (ground pad)) between the date 2010-06-04 and 2017-03-20, in descending order.

**Reasoning**:
To rank the landing outcomes by count within the specified date range, I need to execute a SQL query that filters records by date, groups them by 'Landing_Outcome', counts the occurrences, and then orders these counts in descending order.



In [ ]:
%%sql
SELECT "Landing_Outcome", COUNT("Landing_Outcome") AS Outcome_Count
FROM SPACEXTABLE
WHERE Date BETWEEN '2010-06-04' AND '2017-03-20'
GROUP BY "Landing_Outcome"
ORDER BY Outcome_Count DESC;

 * sqlite:///my_data1.db
Done.


[('No attempt', 10),
 ('Success (drone ship)', 5),
 ('Failure (drone ship)', 5),
 ('Success (ground pad)', 3),
 ('Controlled (ocean)', 3),
 ('Uncontrolled (ocean)', 2),
 ('Failure (parachute)', 2),
 ('Precluded (drone ship)', 1)]

## Summary of Results and Insights

Based on the SQL queries executed, we can summarize the findings from the SpaceX launch data:

*   **Unique Launch Sites**: The unique launch sites are 'CCAFS LC-40', 'VAFB SLC-4E', 'KSC LC-39A', and 'CCAFS SLC-40'.
*   **Launch Sites starting with 'CCA'**: We identified 5 launch records from 'CCAFS LC-40'.
*   **Total Payload Mass by NASA (CRS)**: Boosters launched by NASA (CRS) carried a total payload mass of 45596 kg.
*   **Average Payload Mass for F9 v1.1**: The average payload mass for booster version 'F9 v1.1' was 2928.4 kg.
*   **First Successful Ground Pad Landing**: The first successful landing on a ground pad occurred on '2015-12-22'.
*   **Boosters with Specific Drone Ship Success and Payload**: The booster versions with a 'Success (drone ship)' landing outcome and a payload mass between 4000 kg and 6000 kg are 'F9 FT B1022', 'F9 FT B1026', 'F9 FT B1021.2', and 'F9 FT B1031.2'.
*   **Mission Outcomes Count**: There were 98 'Success' missions, 1 'Failure (in flight)', 1 'Success ', and 1 'Success (payload status unclear)'.
*   **Booster Versions with Maximum Payload Mass**: Multiple booster versions carried the maximum payload mass, including 'F9 B5 B1048.4', 'F9 B5 B1049.4', 'F9 B5 B1051.3', 'F9 B5 B1056.4', 'F9 B5 B1048.5', 'F9 B5 B1051.4', 'F9 B5 B1049.5', 'F9 B5 B1060.2 ', 'F9 B5 B1058.3 ', 'F9 B5 B1051.6', 'F9 B5 B1060.3', and 'F9 B5 B1049.7 '.
*   **2015 Drone Ship Failures**: In 2015, there were two 'Failure (drone ship)' outcomes: one in January with 'F9 v1.1 B1012' from 'CCAFS LC-40' and another in April with 'F9 v1.1 B1015' from 'CCAFS LC-40'.
*   **Ranked Landing Outcomes (2010-06-04 to 2017-03-20)**:
    1.  'No attempt': 10
    2.  'Success (drone ship)': 5
    3.  'Failure (drone ship)': 5
    4.  'Success (ground pad)': 3
    5.  'Controlled (ocean)': 3
    6.  'Uncontrolled (ocean)': 2
    7.  'Failure (parachute)': 2
    8.  'Precluded (drone ship)': 1

These insights provide a comprehensive overview of SpaceX's launch activities, highlighting trends in launch site usage, payload capabilities, and landing success rates over specific periods.

## Final Task

### Subtask:
Summarize the results of all the SQL queries executed and provide insights based on the analysis.


## Summary:

### Q&A

*   **What are the unique launch sites?** The unique launch sites are 'CCAFS LC-40', 'VAFB SLC-4E', 'KSC LC-39A', and 'CCAFS SLC-40'.
*   **How many records have launch sites starting with 'CCA'?** Five records were displayed, all from 'CCAFS LC-40'.
*   **What is the total payload mass for NASA (CRS) missions?** The total payload mass for NASA (CRS) missions is 45596 kg.
*   **What is the average payload mass for the F9 v1.1 booster version?** The average payload mass for 'F9 v1.1' is 2928.4 kg.
*   **When was the first successful ground pad landing?** The first successful ground pad landing occurred on '2015-12-22'.
*   **Which booster names had successful drone ship landings with payload mass between 4000 and 6000 kg?** The boosters are 'F9 FT B1022', 'F9 FT B1026', 'F9 FT B1021.2', and 'F9 FT B1031.2'.
*   **What are the total successful and failure mission outcomes?** There were 98 'Success', 1 'Failure (in flight)', 1 'Success ', and 1 'Success (payload status unclear)' missions.
*   **Which booster versions carried the maximum payload mass?** Multiple booster versions carried the maximum payload mass, including 'F9 B5 B1048.4', 'F9 B5 B1049.4', 'F9 B5 B1051.3', 'F9 B5 B1056.4', 'F9 B5 B1048.5', 'F9 B5 B1051.4', 'F9 B5 B1049.5', 'F9 B5 B1060.2 ', 'F9 B5 B1058.3 ', 'F9 B5 B1051.6', 'F9 B5 B1060.3', and 'F9 B5 B1049.7 '.
*   **What were the month names, booster versions, and launch sites for drone ship failures in 2015?**
    *   **January 2015**: 'F9 v1.1 B1012' from 'CCAFS LC-40' resulted in 'Failure (drone ship)'.
    *   **April 2015**: 'F9 v1.1 B1015' from 'CCAFS LC-40' resulted in 'Failure (drone ship)'.
*   **How are landing outcomes ranked by count between 2010-06-04 and 2017-03-20?**
    1.  'No attempt': 10
    2.  'Success (drone ship)': 5
    3.  'Failure (drone ship)': 5
    4.  'Success (ground pad)': 3
    5.  'Controlled (ocean)': 3
    6.  'Uncontrolled (ocean)': 2
    7.  'Failure (parachute)': 2
    8.  'Precluded (drone ship)': 1

### Data Analysis Key Findings

*   The dataset contains launch data from four unique sites: 'CCAFS LC-40', 'VAFB SLC-4E', 'KSC LC-39A', and 'CCAFS SLC-40'.
*   NASA (CRS) missions are significant clients, accounting for a total payload mass of 45596 kg.
*   The first successful ground pad landing by SpaceX occurred on '2015-12-22', marking a key milestone in their reusable rocket technology development.
*   Specific booster versions like 'F9 FT B1022', 'F9 FT B1026', 'F9 FT B1021.2', and 'F9 FT B1031.2' demonstrated successful drone ship landings with mid-range payloads (4000-6000 kg).
*   A vast majority of missions recorded were successful (98 'Success' out of 101 recorded outcomes), with only one explicit 'Failure (in flight)'.
*   Many advanced 'F9 B5' booster versions consistently achieved the maximum payload capacity, indicating optimized heavy-lift capabilities in later generations.
*   During the period 2010-06-04 to 2017-03-20, 'No attempt' was the most frequent landing outcome (10 times), closely followed by equally common 'Success (drone ship)' and 'Failure (drone ship)' outcomes (5 times each), highlighting a significant experimental phase with drone ship landings.

### Insights or Next Steps

*   The high number of booster versions achieving maximum payload suggests standardization or incremental improvements within the F9 B5 series, implying a mature design for heavy payloads. Further analysis could explore the performance differences (e.g., reusability, cost-efficiency) among these maximum-payload boosters.
*   The identical count of 'Success (drone ship)' and 'Failure (drone ship)' outcomes (5 each) in the early-to-mid period (2010-2017) indicates the experimental nature and learning curve associated with this complex landing method. It would be insightful to analyze the specific factors contributing to these failures versus successes to understand the improvements made over time.
